File: 76. Minimum Window Substring/MinWindow.py
from collections import Counter
from typing import Dict


class Solution:
    # O(m + n) time,
    # O(m) space,
    # Approach: two pointers, sliding window
    def minWindow(self, s: str, t: str) -> str:
        m = len(s)
        n = len(t)
        t_count = Counter(t) # O(1) space,
        
        # O(1) time,
        def areWindowsEqual(t_count: Dict[str: int], window: Dict[str: int]) -> bool:
            for key, val in t_count.items():
                if key in window and window[key] >= val:
                    continue
                return False
            
            return True
        
        
        if n > m:
            return ""
        
        min_len = float('inf')
        result = ''
        window = {} # O(1) space,
        l, r = 0, n
        
        for ch in s[l:r]:
            window[ch] = window.get(ch, 0) + 1
            
        if t_count == window:
            return s[l:r]
        
        while r <= m:
            areEqual = areWindowsEqual(t_count, window)
            if areEqual:
                if min_len > r - l:
                    min_len = (r - l)
                    result = s[l:r]
                window[s[l]] -=1
                l +=1
            else:
                if r == m:
                    break
                window[s[r]] = window.get(s[r], 0) + 1
                r +=1
                
        
        return result


sol = Solution()
print(sol.minWindow("ADOBECODEBANC"
,"ABC"))

File: 76. Minimum Window Substring/min_window.rs
use std::collections::HashMap;


fn are_equal(needle: &HashMap<char, i32>, haystack: &HashMap<char, i32>) -> bool {
    for (k, v) in needle {
        if let Some(count) = haystack.get(k) {
            if *count < *v {
                return false;
            }
        } else {
            return false;
        }
    }
    
    return true;
}


impl Solution {
    
    
    // O(n + m) time, n -> s.len(), m -> t.len(),
    // O(1) space,
    // Approach: sliding window, hashmap
    pub fn min_window(s: String, t: String) -> String {
        let (mut left, mut right) = (0, 0);
        let mut answer = "";
        
        let mut t_dict: HashMap<char, i32> = HashMap::new();
        for ch in t.chars() {
            let mut count = 1;
            if let Some(prev_count) = t_dict.get(&ch) {
                count += *prev_count 
            }
            t_dict.insert(ch, count);
        }
        let mut window: HashMap<char, i32> = HashMap::new();
        let chars: Vec<char> = s.chars().collect();
        
        while right < chars.len() {
            let curr = chars[right];
            let mut curr_count = 0;
            if let Some(count) = window.get(&curr) {
                curr_count = *count;
            }
            window.insert(curr, curr_count+1);
            
            while are_equal(&t_dict, &window) {
                
                // udpate our answer
                let slice = &s[left..right+1];
                if answer == "" || slice.len() < answer.len() {
                    answer = slice;
                }
                
                // udpate our window
                let left_char = chars[left];
                let left_count = window.get(&left_char).unwrap();
                window.insert(left_char, left_count-1);
                left += 1;
            }
            
            right += 1;
        }
        
        answer.to_string()
    }
}